### PR TITLE
Fix compute issues

### PR DESCRIPTION
**Pull request by rodrigogiraoserrao**
**Please review the following checklist.**

- [x] Docstrings on all new or modified functions / classes 
- [ ] Updated documentation
- [ ] Updated CHANGELOG.md (where appropriate)

### CODE REVIEW

**Comment by rodrigogiraoserrao** [comment]
As it stands, the only commit here is an attempt at fixing #1218 which it does – partially. The `xfail` test is now passing but now I broke a test

```
pytest tests/snapshot_tests/test_snapshots.py::test_order_independence -vv
```

Locally, running the app produces the correct output and we can "fix" the test if we start by pausing 50ms before pressing `"t"`.

However, there _must_ be an issue because now we can't even run the following example:

```
textual run --dev docs/examples/guide/reactivity/computed01.py
```

What is seems is that now I am being too eager in calling compute methods.

As of writing this, my changes make it so that **whenever** a reactive attribute changes, we go over **all** compute methods and try to call them. If they are synchronous, they are called immediately. If they are async, we post a Callback event to call them later. We do this after checking the watchers associated with the attribute.

We do not check all compute methods again after a `watch` method is done running because if the `watch` method changed any reactive attribute, the compute methods will react to that change in time.

This sounds like the strategy that should be followed but it looks like it doesn't work very well during app startup.

**Comment by rodrigogiraoserrao** [comment]
1. A potential fix would entail having the user tag compute methods with the reactive attributes it depends on. Then:
    - we would know exactly what compute methods need to be called when a specific reactive attribute changes; and
    - we could ensure all attribute dependencies have been initialised before running a compute method.
 2. Another potential fix would be to try and initialise all reactive attributes of an object before running compute methods for the first time.

**Comment by rodrigogiraoserrao** [comment]
Plot twist, I was trying to fix #1218 but ended up fixing #1217...

**Comment by rodrigogiraoserrao** [comment]
Plot twist twist: synchronous compute methods really are being called immediately, which is the object of issue 1218. However, the expected output of the `xfail` test for issue 1218 doesn't match the output we are getting because we call compute functions more often that what seems to have been assumed when the original test was written.

In particular, as of this comment, every time a reactive attribute changes, we trigger two compute calls:

 1. the first one, that is expected; but then
 2. the compute method finishes, we set the value of the computed attribute, and that triggers a new wave of `compute` calls

File: src/textual/reactive.py [hunk 1]
```diff
@@ -182,15 +182,47 @@ def __set__(self, obj: Reactable, value: ReactiveType) -> None:
...
+    @classmethod
+    def _check_computes(cls, obj: Reactable) -> None:
+        """Invoke all computes.
+
+        Args:
+            obj (Reactable): Reactable object whose computes will be called.
+        """
+        _rich_traceback_guard = True
+
+        async def _set_later(obj: Reactable, compute: str, compute_method: Callable) -> None:
```
Comment (willmcgugan) [inline src/textual/reactive.py:202]: Docstrings please. Even on private methods and closures.

**Review by willmcgugan — APPROVED** [review]

**Comment by willmcgugan** [comment]
@rodrigogiraoserrao I think there are a number of interactions going on here.

The PR is good, but I think I need to review the reactive stuff. I have a feeling that its like a rubic cube, where each PR is solving one side only.

Let's put this on hold until I've done that.

**Comment by willmcgugan** [comment]
On the right track, but there were larger issues that needed to be addressed at the same time! Closed in favor of https://github.com/Textualize/textual/pull/1558

**Pull request closed without merging**